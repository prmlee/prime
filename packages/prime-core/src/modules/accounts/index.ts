import { AccountsModule } from '@accounts/graphql-api';
import { AccountsPassword } from '@accounts/password';
import { AccountsServer } from '@accounts/server';
import { AccountsTypeorm } from '@accounts/typeorm';
import Mailgun from 'mailgun-js';
import { Connection } from 'typeorm';

const { MAILGUN_API_KEY, MAILGUN_DOMAIN, ACCOUNTS_SECRET } = process.env;

let mailgun;

if (MAILGUN_API_KEY && MAILGUN_API_KEY !== '') {
  mailgun = Mailgun({
    apiKey: MAILGUN_API_KEY,
    domain: MAILGUN_DOMAIN,
  });
}

export const createAccounts = async (connection: Connection) => {
  const tokenSecret = ACCOUNTS_SECRET || 'not very secret secret';
  const db = new AccountsTypeorm({ connection, cache: 1000 });
  const password = new AccountsPassword({
    twoFactor: {
      appName: 'Prime',
    },
  });
  const accountsServer = new AccountsServer(
    {
      db,
      tokenSecret,
      siteUrl: 'http://localhost:3000',
      sendMail(mail) {
        if (mailgun) {
          return mailgun.messages().send(mail);
        }
      },
    },
    { password }
  );

  const accounts = AccountsModule.forRoot({
    accountsServer,
    headerName: 'x-prime-token',
  });

  return accounts;
};
