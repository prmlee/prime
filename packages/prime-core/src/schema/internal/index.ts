import { GraphQLModule } from '@graphql-modules/core';
import { PubSub } from 'apollo-server-express';
import { isNumber, omitBy } from 'lodash';
import { Connection } from 'typeorm';
import { buildTypeDefsAndResolvers } from '../../utils/build-resolvers/buildTypeDefsAndResolvers';
import { WebhookResolver } from './resolvers/WebhookResolver';
export const pubSub = new PubSub();

export const createInternal = async (connection: Connection) => {
  const schema = await buildTypeDefsAndResolvers({
    resolvers: [WebhookResolver],
    pubSub,
  });

  return new GraphQLModule({
    name: 'Prime.Internal',
    typeDefs: () => [schema.typeDefs],
    resolvers: ({ config = {} }) =>
      omitBy(schema.resolvers, (item: any, key: string) => {
        if (key === 'User') {
          return true;
        }
        if (typeof item === 'object' && Object.values(item).every(isNumber)) {
          return true;
        }
        return false;
      }),
    configRequired: false,
  });
};
