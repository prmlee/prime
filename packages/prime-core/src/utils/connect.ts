import { createConnection } from 'typeorm';
import { PostgresDriver } from 'typeorm/driver/postgres/PostgresDriver';

export const connect = (url = process.env.DATABASE_URL) =>
  createConnection({
    type: 'postgres',
    url,
    entities: [...require('@accounts/typeorm').entities, 'src/entities/*.ts'],
    synchronize: true,
    logger: 'debug',
  }).then(connection => {
    const driver = connection.driver as PostgresDriver;

    // Fixes postgres timezone bug
    driver.postgres.defaults.parseInputDatesAsUTC = true;
    driver.postgres.types.setTypeParser(1114, (str: any) => new Date(str + 'Z'));

    return connection;
  });
