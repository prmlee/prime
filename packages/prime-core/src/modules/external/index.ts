import { AccountsModule } from '@accounts/graphql-api';
import AccountsServer from '@accounts/server';
import { GraphQLModule } from '@graphql-modules/core';
import { PrimeFieldOperation } from '@primecms/field';
import { AuthenticationError } from 'apollo-server-core';
import debug from 'debug';
import {
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  GraphQLUnionType,
  printSchema,
} from 'graphql';
import { camelCase, omit, upperFirst } from 'lodash';
import { createResolversMap } from 'type-graphql/dist/utils/createResolversMap';
import Container from 'typedi';
import { Connection, getRepository } from 'typeorm';
import { Document } from '../../entities/Document';
import { Schema, SchemaVariant } from '../../entities/Schema';
import { DocumentTransformer } from '../../utils/DocumentTransformer';
import { createAllDocumentResolver } from './resolvers/createAllDocumentResolver';
import { createDocumentCreateResolver } from './resolvers/createDocumentCreateResolver';
import { createDocumentRemoveResolver } from './resolvers/createDocumentRemoveResolver';
import { createDocumentResolver } from './resolvers/createDocumentResolver';
import { createDocumentUpdateResolver } from './resolvers/createDocumentUpdateResolver';
import { documentUnionResolver } from './resolvers/documentUnionResolver';
import { createSchemaConnectionType } from './types/createSchemaConnectionType';
import { createSchemaInputType } from './types/createSchemaInputType';
import { createSchemaType } from './types/createSchemaType';
import { DocumentRemove } from './types/DocumentRemove';
import { resetTypeNames, uniqueTypeName } from './utils/uniqueTypeNames';

export const log = debug('prime:graphql');

export const getDefaultLocale = () => 'en';

export const createExternal = async (connection: Connection) => {
  log('building schema');
  resetTypeNames();

  const documentTransformer = new DocumentTransformer();
  const schemas = await getRepository(Schema).find();

  const types = new Map();
  const queries: { [key: string]: any } = {};
  const mutations: { [key: string]: any } = {};
  const resolvers: { [key: string]: any } = {};

  for (const schema of schemas) {
    if (schema.variant === SchemaVariant.Template) {
      continue;
    }

    if (schema.variant === SchemaVariant.Slice) {
      // maybe do something special?
      schema.name = uniqueTypeName(upperFirst(camelCase(schema.name)));
    } else {
      schema.name = uniqueTypeName(upperFirst(camelCase(schema.name)));
    }

    schema.fields = await documentTransformer.getFields(schema);

    const { name, fields } = schema;
    const payload = { schema, schemas, fields, name, resolvers, types, documentTransformer };
    const SchemaTypeConfig = await createSchemaType(payload);

    const { CREATE, UPDATE } = PrimeFieldOperation;
    const SchemaType = SchemaTypeConfig.type;

    types.set(name, SchemaTypeConfig);

    const connectionType = await createSchemaConnectionType(payload, SchemaType);
    const createType = await createSchemaInputType(payload, SchemaType, CREATE);
    const updateType = await createSchemaInputType(payload, SchemaType, UPDATE);

    if (schema.variant === SchemaVariant.Default) {
      queries[name] = SchemaTypeConfig;
      queries[`all${name}`] = connectionType;
      mutations[`create${name}`] = createType;
      mutations[`update${name}`] = updateType;
      mutations[`remove${name}`] = DocumentRemove;
    }

    types.set(`create${name}`, createType);
    types.set(`update${name}`, updateType);
  }

  for (const [, type] of types) {
    const { asyncResolve } = type || { asyncResolve: null };
    if (asyncResolve) {
      await asyncResolve();
    }
  }

  for (const schema of schemas) {
    if (!types.has(schema.name) || schema.variant !== SchemaVariant.Default) {
      continue;
    }

    const { name, fields } = schema;
    const payload = { schema, schemas, fields, name, resolvers, types, documentTransformer };

    resolvers[name] = await createDocumentResolver(payload);
    resolvers[`all${name}`] = await createAllDocumentResolver(payload);
    resolvers[`create${name}`] = await createDocumentCreateResolver(payload);
    resolvers[`update${name}`] = await createDocumentUpdateResolver(payload);
    resolvers[`remove${name}`] = await createDocumentRemoveResolver(payload);

    queries[name].resolve = resolvers[name];
    queries[`all${name}`].resolve = resolvers[`all${name}`];
    mutations[`create${name}`].resolve = resolvers[`create${name}`];
    mutations[`update${name}`].resolve = resolvers[`update${name}`];
    mutations[`remove${name}`].resolve = resolvers[`remove${name}`];

    const SchemaType = types.get(schema.name)!.type!;

    if (!SchemaType || Object.keys(omit(SchemaType.getFields(), ['id', '_meta'])).length === 0) {
      delete queries[name];
      delete queries[`all${name}`];
      delete mutations[`create${name}`];
      delete mutations[`update${name}`];
      delete mutations[`remove${name}`];
      types.delete(name);
    }
  }

  const primeDocumentNotFoundTypeName = uniqueTypeName('Prime_Document_NotFound');
  const PrimeDocumentNotFound = new GraphQLObjectType({
    name: primeDocumentNotFoundTypeName,
    fields: { message: { type: GraphQLString } },
  });
  types.set(primeDocumentNotFoundTypeName, { type: PrimeDocumentNotFound });

  const primeDocumentTypeName = uniqueTypeName('Prime_Document');
  queries[primeDocumentTypeName] = {
    args: {
      id: { type: GraphQLString },
      locale: { type: GraphQLString },
    },
    type: new GraphQLUnionType({
      name: primeDocumentTypeName,
      types: [
        ...Array.from(types.values())
          .filter(
            ({ variant, operation }) =>
              variant === SchemaVariant.Default && operation === PrimeFieldOperation.READ
          )
          .map(typeConfig => typeConfig.type),
        PrimeDocumentNotFound,
      ],
    }),
    resolve: documentUnionResolver(resolvers),
  };

  const hasMutations = Object.keys(mutations).length;

  const graphqlSchema = new GraphQLSchema({
    query: new GraphQLObjectType({
      name: 'Query',
      fields: queries,
    }),
    ...(hasMutations && {
      mutation: new GraphQLObjectType({
        name: 'Mutation',
        fields: mutations,
      }),
    }),
  });

  const resolverMap = createResolversMap(graphqlSchema);
  const unionResolvers = Object.entries(resolverMap).reduce((acc, item) => {
    const [key, value] = item as any;
    if (key && value.__resolveType) {
      acc[key] = value;
    }
    return acc;
  }, {});

  const PRIME_TOKEN = 'x-prime-token';
  const PRIME_PREVIEW = 'x-prime-preview';

  return new GraphQLModule({
    name: 'prime-graphql',
    extraSchemas: [graphqlSchema],
    resolvers: unionResolvers,
    async context({ req }) {
      const requestId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
      log('requestId', requestId);
      const container = Container.of(requestId);
      const ctx: { [key: string]: any } = {
        requestId,
        container,
      };

      if (req.headers[PRIME_TOKEN]) {
        const server = AccountsModule.injector.get(AccountsServer);
        ctx.userSession = await server.findSessionByAccessToken(req.headers[PRIME_TOKEN]);
        if (!ctx.userSession) {
          throw new AuthenticationError('Token not valid');
        }

        if (req.headers[PRIME_PREVIEW]) {
          ctx.preview = await getRepository(Document).findOneOrFail(req.headers[PRIME_PREVIEW]);
        }
      }

      container.set('context', ctx);
      return ctx;
    },
    logger: {
      clientError: () => null,
      log,
      error: log,
    },
  });
};
