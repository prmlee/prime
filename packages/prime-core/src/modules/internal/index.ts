import { GraphQLModule } from '@graphql-modules/core';
import debug from 'debug';
import { mapValues, omitBy } from 'lodash';
import { buildTypeDefsAndResolvers } from 'type-graphql';
import Container from 'typedi';
import { Connection } from 'typeorm';
import { pubSub } from '../../utils/pubSub';
import { createAccounts } from '../accounts';
import { AccessTokenResolver } from './resolvers/AccessTokenResolver';
import { DocumentResolver } from './resolvers/DocumentResolver';
import { PrimeResolver } from './resolvers/PrimeResolver';
import { ReleaseResolver } from './resolvers/ReleaseResolver';
import { SchemaResolver } from './resolvers/SchemaResolver';
import { UserResolver } from './resolvers/UserResolver';
import { WebhookResolver } from './resolvers/WebhookResolver';
import { abilityForbiddenMiddleware } from './utils/abilityErrorMiddleware';
import { createAbility } from './utils/createAbility';
import { isAuthenticated } from './utils/isAuthenticated';
import { noEnumsOrInheritedModels } from './utils/noEnumsOrInheritedModels';
import { noUndefinedTypeOf } from './utils/noUndefinedTypeOf';

export const log = debug('prime:core');

export const createInternal = async (connection: Connection) => {
  log('building schema');

  const accounts = await createAccounts(connection);

  const schema = await buildTypeDefsAndResolvers({
    resolvers: [
      AccessTokenResolver,
      DocumentResolver,
      PrimeResolver,
      ReleaseResolver,
      SchemaResolver,
      UserResolver,
      WebhookResolver,
    ],
    pubSub,
    globalMiddlewares: [abilityForbiddenMiddleware],
    container: ({ context }) => context.container,
  });

  return new GraphQLModule({
    name: 'prime-core',
    imports: [accounts],
    typeDefs: () => [schema.typeDefs],
    resolvers: () =>
      mapValues(omitBy(schema.resolvers, noEnumsOrInheritedModels), noUndefinedTypeOf),
    async context(session, currentContext) {
      const requestId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
      const container = Container.of(requestId);

      const ctx = {
        requestId,
        container,
        ability: createAbility(currentContext),
      };

      container.set('context', ctx);
      return ctx;
    },
    resolversComposition: {
      'Mutation.createUser': [isAuthenticated()],
    },
    configRequired: false,
    logger: {
      clientError: () => null,
      log,
      error: log,
    },
  });
};
