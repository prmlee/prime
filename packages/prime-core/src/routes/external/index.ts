import { ApolloServer } from 'apollo-server-express';
import { createContext } from 'dataloader-sequelize';
import * as express from 'express';
import { GraphQLBoolean, GraphQLID, GraphQLObjectType, GraphQLSchema } from 'graphql';
import { get, omit } from 'lodash';
import { ContentType } from '../../models/ContentType';
import { ContentTypeField } from '../../models/ContentTypeField';
import { sequelize } from '../../sequelize';
import { contentEntryMetaType } from '../../types/contentEntryMetaType';
import { create } from './create';
import { find } from './find';
import { findAll } from './findAll';
import { remove } from './remove';
import { update } from './update';
import { resolveFieldType } from './utils/resolveFieldType';
import { EntryTransformer } from '../../utils/entryTransformer';
import { Sentry } from '../../utils/Sentry';
import { Settings } from '../../models/Settings';
import { ContentEntry } from '../../models/ContentEntry';

// import { User } from '../../models/User';
// import { acl } from '../../acl';

interface IContext {
  settings: any;
  sequelizeDataLoader: any;
  public: boolean;
  published: null | boolean;
  contentReleaseId: null | string;
  versionId: null | string;
  entryId: null | string;
};

export const entryTransformer = new EntryTransformer();

// tslint:disable-next-line no-require-imports no-var-requires
export const debug = require('debug')('prime:graphql');

// tslint:disable-next-line max-func-body-length
export const externalGraphql = async () => {

  const app = express();

  entryTransformer.resetTransformCache();

  const queries = {
    __slices: {},
  };
  const inputs = {};
  const models = {
    ContentEntry,
    ContentType,
  };

  const settings = await Settings.get();

  const contentTypes = await ContentType.findAll();

  debug(contentTypes.length, 'content types');

  await Promise.all(
    contentTypes.map(async (contentType) => {
      const contentTypeIds = get(contentType, 'settings.contentTypeIds', []);
      contentTypeIds.push(contentType.id);

      const fields = await ContentTypeField.findAll({
        where: {
          contentTypeId: contentTypeIds
        }
      });

      contentType.fields = [
        ...fields.filter(field => field.contentTypeId === contentType.id),
        ...fields.filter(field => field.contentTypeId !== contentType.id),
      ];

      entryTransformer.cache.set(contentType.id, contentType.fields);
    })
  );

  // slices are prio
  contentTypes.sort((a: any, b: any) => b.isSlice - a.isSlice);

  await Promise.all(
    contentTypes.map(async (contentType) => {
      const outputFields = () => contentType.fields.reduce(
        (acc, field: ContentTypeField) => {
          const fieldType = resolveFieldType(field);
          if (fieldType) {
            acc[field.name] = fieldType.getGraphQLOutput({
              field,
              queries,
              models,
              contentType,
              contentTypes,
              resolveFieldType
            });
          }
          if (!acc[field.name]) {
            delete acc[field.name];
          }

          return acc;
        },
        {}
      );

      // tslint:disable-next-line variable-name
      const GraphQLContentType = new GraphQLObjectType({
        name: contentType.name,
        fields: () => ({
          id: { type: GraphQLID },
          ...outputFields(),
          _meta: { type: contentEntryMetaType }
        })
      });

      if (contentType.isSlice) {
        debug('content type %s is a slice', contentType.id);
        queries.__slices[contentType.name] = {
          outputType: new GraphQLObjectType({
            name: `${contentType.name}Slice`,
            fields: outputFields,
          }),
          inputType: null, // @todo we maybe want to store this type for further reuse
        }
        return null;
      }

      if (contentType.isTemplate) {
        debug('content type %s is a template', contentType.id);

        return null;
      }

      debug(
        'created content type:',
        `${contentType.name}(${contentType.fields.map(field => field.name).join(', ')})`
      );

      queries[contentType.name]             = find({ GraphQLContentType, contentType, contentTypes, queries });
      queries[`all${contentType.name}`]  = findAll({ GraphQLContentType, contentType, contentTypes, queries });

      if (get(contentType, 'settings.mutations', true) === true) {
        inputs[`create${contentType.name}`] = create({ GraphQLContentType, contentType, contentTypes, queries });
        inputs[`update${contentType.name}`] = update({ GraphQLContentType, contentType, contentTypes, queries });
        inputs[`remove${contentType.name}`] = remove({ GraphQLContentType, contentType, contentTypes, queries });
      }
    })
  );

  const queriesAndMutations: any = {}; // tslint:disable-line no-any
  const realQueries = omit(queries, '__slices') as any;

  if (Object.keys(realQueries).length === 0) {
    realQueries[`isEmpty`] = {
      type: GraphQLBoolean,
      resolve() {
        return true;
      }
    };
  } else if (Object.keys(inputs).length > 0) {
    queriesAndMutations.mutation = new GraphQLObjectType({
      name: 'Mutation',
      fields: inputs
    });
  }

  queriesAndMutations.query = new GraphQLObjectType({
    name: 'Query',
    fields: realQueries
  });

  const schema = new GraphQLSchema(queriesAndMutations);

  const server = new ApolloServer({
    introspection: true,
    tracing: true,
    schema,
    context: async ({ req }) => {
      try {
        const cookie = String(req.headers.cookie);
        const cookies = new Map(cookie.split(';').map(n => n.trim().split('=')) as any);

        if (req.headers['x-prime-version-id'] && req.headers['x-prime-version-id'].length === 36) {
          cookies.set('prime.versionId', req.headers['x-prime-version-id']);
        }

        const context: IContext = {
          settings,
          sequelizeDataLoader: createContext(sequelize),
          public: (settings.accessType === 'public'),
          published: true,
          contentReleaseId: null,
          versionId: null,
          entryId: null,
        };

        debug('context.public %o', context.public);

        if (cookies.has('prime.versionId')) {
          const versionId = cookies.get('prime.versionId') as string;
          debug('context.versionId %o', versionId);
          if (versionId.length === 36) {
            const entry = await ContentEntry.findOne({ where: { versionId }});
            if (entry) {
              context.entryId = entry.id;
              if (entry.contentReleaseId) {
                context.contentReleaseId = entry.contentReleaseId;
              } else if (!entry.isPublished) {
                context.published = null;
              }
            } else {
              context.contentReleaseId = versionId;
              context.published = null;
            }

            context.versionId = versionId;

            debug('context.published %o', context.published);
            debug('context.contentReleaseId %o', context.contentReleaseId);
          }
        }

        if (req.user && String(req.headers.referer).match(/\/graphql(\/?\?.*)?$/) && !req.headers['x-prime-published']) {
          context.published = null;
          debug('context.published %o %s', context.published, '(overwrite)');
        }

        return context;
      } catch (err) {
        console.error(err);
        return {};
      }
    },
    formatError(error) {
      if (Sentry) {
        Sentry.captureException(error);
      }

      return error;
    }
  });

  server.applyMiddleware({
    app,
    cors: {
      origin: true,
    }
  });

  return app;
};
