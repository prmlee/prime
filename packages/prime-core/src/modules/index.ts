import { GraphQLModule } from '@graphql-modules/core';
import Container from 'typedi';
import { Connection } from 'typeorm';
import { createAccounts } from './accounts';
import { createInternal } from './internal';

export const createModules = async (connection: Connection) => {
  const accounts = await createAccounts(connection);
  const internal = await createInternal(connection);

  return new GraphQLModule({
    imports: () => [internal, accounts],
    context() {
      const requestId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
      const container = Container.of(requestId);
      const ctx = {
        requestId,
        container,
        user: { id: '35416EC9-CA8D-4B7F-B36C-C16C1E88B255'.toLowerCase() }, //
      };
      container.set('context', ctx);
      return ctx;
    },
  });
};
