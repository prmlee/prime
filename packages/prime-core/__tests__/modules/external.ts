import { User } from '@accounts/typeorm';
import { GraphQLModule } from '@graphql-modules/core';
import { gql } from 'apollo-server-core';
import { execute } from 'graphql';
import 'reflect-metadata';
import Container from 'typedi';
import { Connection, getRepository, useContainer } from 'typeorm';
import { createExternal } from '../../src/modules/external';
import { createInternal } from '../../src/modules/internal';
import { connect } from '../../src/utils/connect';
import { createSchemaQuery } from '../utils/createSchemaQuery';

useContainer(Container);

describe('InternalModule', () => {
  let connection: Connection;
  let external: GraphQLModule;
  let internal: GraphQLModule;
  let user;
  const context: any = { req: { headers: {} } };

  const query = async (document, which = external, ctx = context) => {
    const contextValue = await which.context(ctx);
    contextValue.user = user;
    return execute({
      schema: which.schema,
      contextValue,
      document,
    });
  };

  beforeAll(async () => {
    connection = await connect(process.env.TEST_DATABASE_URL);

    await connection.dropDatabase();
    await connection.synchronize();

    internal = await createInternal(connection);
    external = await createExternal(connection);

    const onboard = await query(
      gql`
        mutation {
          onboard(email: "demodemo@gmail.com", password: "demodemo")
        }
      `,
      internal
    );
    expect(onboard!.data!.onboard).toBe(true);
    user = await getRepository(User).findOne();
  });

  afterAll(() => connection.close());

  describe('GraphQL Module', () => {
    it('should have schema', () => {
      expect(external.schema).toBeTruthy();
    });
    it('should have resolvers', () => {
      expect(external.resolvers).toBeTruthy();
    });
  });

  describe('health', () => {
    it('should have fields', async () => {
      const result = await query(
        gql`
          query {
            allFields {
              type
              title
              description
              options
              ui
            }
          }
        `,
        internal
      );
      const fields = result.data!.allFields!.map(({ ui, ...rest }) => rest);
      expect(fields).toMatchSnapshot();
    });
  });

  describe('No Schemas', () => {
    it('should work to query PrimeDocument', async () => {
      const result = await query(gql`
        query {
          Prime_Document {
            ... on Prime_Document_NotFound {
              message
            }
          }
        }
      `);
      expect(result).toEqual({ data: { Prime_Document: null } });
    });
  });

  describe('One Schema', () => {
    beforeAll(async () => {
      const fields = [{ name: 'name' }, { name: 'bio' }];
      await query(
        gql`
          mutation {
            createSchema(
              input: ${createSchemaQuery('Author', fields)}
            ) {
              id
              title
              name
            }
          }
        `,
        internal
      );
      external = await createExternal(connection);
    });

    it('should have Author schema', async () => {
      const res = await query(
        gql`
          query {
            __schema {
              types {
                name
              }
            }
          }
        `
      );
      expect(res!.data!.__schema!.types.find(n => n.name === 'Author')).toEqual({ name: 'Author' });
    });

    it('should be able to create author', async () => {
      const res = await query(
        gql`
          mutation {
            createAuthor(input: { name: "John Doe", bio: "Anonymous person" }) {
              id
              _meta {
                id
              }
            }
          }
        `
      );

      const {
        id,
        _meta: { id: uuid },
      } = res!.data!.createAuthor as any;

      expect(id).toBeTruthy();

      const res2 = await query(
        gql`
          query {
            Document(id: "${id}") {
              id
              name
              bio
            }
          }
        `
      );
      expect(res2!.data).toEqual({});

      const publish = await query(
        gql`
          mutation {
            publishDocument(id: "${uuid}") {
              id
            }
          }
      `,
        internal
      );
      expect(publish).toBeTruthy();

      const res3 = await query(
        gql`
          query {
            Author(id: "${id}") {
              id
              name
              bio
            }
          }
        `
      );
      expect(res3!.data!.Author).toEqual({ id, name: 'John Doe', bio: 'Anonymous person' });
    });
  });
});
