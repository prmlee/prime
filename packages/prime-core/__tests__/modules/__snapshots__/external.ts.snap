// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalModule health should have fields 1`] = `
Array [
  Object {
    "description": "Images, videos and other assets",
    "options": Object {
      "crops": Array [],
    },
    "title": "Asset",
    "type": "asset",
  },
  Object {
    "description": "Boolean field",
    "options": Object {
      "label": "",
    },
    "title": "Boolean",
    "type": "boolean",
  },
  Object {
    "description": "Date and time field",
    "options": Object {
      "time": true,
    },
    "title": "DateTime",
    "type": "datetime",
  },
  Object {
    "description": "Link and resolve documents",
    "options": Object {
      "multiple": false,
      "schemaId": null,
      "schemaIds": Array [],
    },
    "title": "Document",
    "type": "document",
  },
  Object {
    "description": "Group other fields to list",
    "options": Object {
      "repeated": true,
    },
    "title": "Group",
    "type": "group",
  },
  Object {
    "description": "Number field",
    "options": Object {
      "float": true,
      "rules": Object {},
    },
    "title": "Number",
    "type": "number",
  },
  Object {
    "description": "Select field",
    "options": Object {
      "enum": false,
      "items": Array [],
      "multiple": false,
      "required": false,
    },
    "title": "Select",
    "type": "select",
  },
  Object {
    "description": "Slice field",
    "options": Object {
      "multiple": true,
      "schemaIds": Array [],
    },
    "title": "Slice",
    "type": "slice",
  },
  Object {
    "description": "Text field with no formatting",
    "options": Object {
      "rules": Object {},
      "type": "singleline",
    },
    "title": "String",
    "type": "string",
  },
]
`;
